components/org.wso2.carbon.identity.application.authenticator.adapter/src/main/java/org/wso2/carbon/identity/application/authenticator/adapter/AbstractAuthenticatorAdapter.java

```diff
@@ -142,14 +166,21 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
     protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
                                                  AuthenticationContext context) throws AuthenticationFailedException {
 
+        /* This method is invoked from the AbstractApplicationAuthenticator class to handle the authentication response.
+         When the ActionExecutionStatus is SUCCESS, the authenticated user is built based on the action response in the
+         processSuccessResponse method within the AuthenticationResponseProcessor class. In this method, if the
+         ActionExecutionStatus is FAILED, an InvalidCredentialsException is thrown, and if the ActionExecutionStatus is
+         ERROR, an AuthenticationFailedException is thrown. Based on the exception, the process method of the superclass
+         updates the authentication context. */
         ActionExecutionStatus executionStatus = (ActionExecutionStatus)
                 context.getProperty(AuthenticatorAdapterConstants.EXECUTION_STATUS_PROP_NAME);
-        if (executionStatus.getStatus() == ActionExecutionStatus.Status.FAILED ||
-                executionStatus.getStatus() == ActionExecutionStatus.Status.ERROR) {
-            /* TODO: Improve AuthenticationFailedException error messages and description with specific error content
-                from the authentication action execution. */
-            throw new AuthenticationFailedException("An error occurred while authenticating with user the " +
-                    " external authentication authentication service.");
+        if (executionStatus.getStatus() == ActionExecutionStatus.Status.FAILED) {
+            throw new InvalidCredentialsException("Authentication for the user with external service failed.",
+                    ((FailedStatus) executionStatus).getResponse().getFailureDescription());
+        } else if (executionStatus.getStatus() == ActionExecutionStatus.Status.ERROR) {
+            setErrorContextForClient(context);
+            throw new AuthenticationFailedException(
```

Comment:
Add a message and a description. Fo description it makes sense to use the description from the external service